syntax = "proto3";

// This is an ugly name. Come up with something better.
package std_daq_protocol;

message GFImageMetadata {
  uint32 scan_id=1;
  uint32 scan_time=2;
  uint32 sync_time=3;
  uint64 frame_timestamp=4;
  uint64 exposure_time=5;
}

message JFImageMetadata {
  uint64 daq_rec=1;
}

message EGImageMetadata {
  uint64 exptime=1;
}

enum ImageMetadataDtype {
  unknown=0;
  uint8=1; uint16=2; uint32=4; uint64=8;
  int8=11; int16=12; int32=14; int64=18;
  float16=22; float32=24; float64=28;
};

enum ImageMetadataStatus {
  good_image=0;
  missing_packets=1;
  id_missmatch=2;
};

message ImageMetadata {
  uint64 image_id=1;
  uint64 height=2;
  uint64 width=3;

  ImageMetadataDtype dtype=4;
  ImageMetadataStatus status=5;

  oneof image_metadata {
    GFImageMetadata gf = 6;
    JFImageMetadata jf = 7;
    EGImageMetadata eg = 8;
  }
}

enum CommandType {
  WRITE_IMAGE = 0;
  START_WRITING = 1;
  STOP_WRITING = 2;
}

message RunInfo {
  uint64 run_id = 1;
  string output_file = 2;
  int32 n_images = 3;
}

message WriterCommand {
  CommandType command_type = 1;
  RunInfo run_info = 2;

  ImageMetadata metadata=3;
}
